### PR TITLE
Raven assembler

### LINKED ISSUE #1042: new module: raven assembler
I think it would be good to have a module for [raven assembler](https://github.com/lbcb-sci/raven)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by fmalmeida**
## PR checklist

Closes #1042 

- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/raven/main.nf [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    raven \\
+        -t $task.cpus \\
+        --graphical-fragment-assembly ${prefix}.gfa \\
+        $options.args \\
+        $reads | \\
+        gzip -c > ${prefix}.fasta.gz
+
```
Comment (mjmansfi) [inline modules/raven/main.nf:37]: The linting complains about trailing whitespace here. Any ideas?

**Review by mjmansfi — COMMENTED** [review]
Hmm, the actions are complaining about linting, editing, and `conda`. The `conda` testing is complaining about a missing dependency - weird, since presumably the containers are based on the `conda` recipe...

```
❯ nf-core modules create-test-yml  --force raven

                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 2.2.dev0



INFO     Press enter to use default values (shown in brackets) or type your own responses                                                                               test_yml_builder.py:54
Test YAML output path (- for stdout) (tests/modules/raven/test.yml): 
INFO     Looking for test workflow entry points: 'tests/modules/raven/main.nf'                                                                                         test_yml_builder.py:118
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
INFO     Building test meta for entry point 'test_raven'                                                                                                               test_yml_builder.py:152
Test name (raven test_raven): 
Test command (nextflow run tests/modules/raven -entry test_raven -c tests/config/nextflow.config): 
Test tags (comma separated) (raven): 
Test output folder with results (leave blank to run test): 
? Choose software profile Conda
INFO     Setting env var '$PROFILE' to 'conda'                                                                                                                         test_yml_builder.py:278
INFO     Running 'raven' test with command:                                                                                                                            test_yml_builder.py:286
         nextflow run tests/modules/raven -entry test_raven -c tests/config/nextflow.config --outdir /tmp/tmp160ftx7y -work-dir /tmp/tmpohqk13c7                                              
CRITICAL Error running test workflow (exit code 1)                                                                                                                             __main__.py:547
         N E X T F L O W  ~  version 21.10.1                                                                                                                                                  
         Launching `tests/modules/raven/main.nf`  - revision: 921c39cb34                                                                                                                      
         [-        ] process > test_raven:RAVEN -                                                                                                                                             
                                                                                                                                                                                              
         [-        ] process > test_raven:RAVEN -                                                                                                                                             
         Creating env using conda: bioconda::raven-assembler=1.6.1                                                                                                                            
                                                                                                                                                                                              
         [-        ] process > test_raven:RAVEN -                                                                                                                                             
         Creating env using conda: bioconda::raven-assembler=1.6.1                                                                                                                            
         Error executing process > 'test_raven:RAVEN (test)'                                                                                                                                  
                                                                                                                                                                                              
         Caused by:                                                                                                                                                                           
           Failed to create Conda environment                                                                                                                                                 
           command: conda create --mkdir --yes --quiet --prefix /tmp/tmpohqk13c7/conda/env-35405f02cf5f78e46e49a318297f0f3b bioconda::raven-assembler=1.6.1                                   
           status : 1                                                                                                                                                                         
           message:                                                                                                                                                                           
             PackagesNotFoundError: The following packages are not available from current channels:                                                                                           
                                                                                                                                                                                              
               - libstdcxx-ng                                                                                                                                                                 
                                                                                                                                                                                              
             Current channels:                                                                                                                                                                
                                                                                                                                                                                              
               - https://repo.anaconda.com/pkgs/main/linux-64                                                                                                                                 
               - https://repo.anaconda.com/pkgs/main/noarch                                                                                                                                   
               - https://repo.anaconda.com/pkgs/r/linux-64                                                                                                                                    
               - https://repo.anaconda.com/pkgs/r/noarch                                                                                                                                      
                                                                                                                                                                                              
             To search for alternate channels that may provide the conda package you're                                                                                                       
             looking for, navigate to                                                                                                                                                         
                                                                                                                                                                                              
                 https://anaconda.org                                                                                                                                                         
                                                                                                                                                                                              
             and use the search bar at the top of the page.  
```

So, not sure what to do here, and I don't know anything about the linting or the editor complaints.

`docker` and `singularity` work for me locally so I think the module works well. Although I know there are more impending changes to module structure that may delay the merge...! :sweat_smile:

**Comment by fmalmeida** [comment]
Hi @mjmansfi,

About the conda error:

```bash
message:                                                                                                                                                                           
             PackagesNotFoundError: The following packages are not available from current channels:                                                                                           
                                                                                                                                                                                              
               - libstdcxx-ng                                                                                                                                                                 
                                                                                                                                                                                              
             Current channels:
```

I suppose you're running in a non-linux machine, right? Because this package is available in conda and in raven recipe, however only for linux: https://anaconda.org/conda-forge/libstdcxx-ng

File: modules/raven/main.nf (hunk 1 above)
Comment (fmalmeida) [reply modules/raven/main.nf:37]: Well, I really don't know. This is my first PR to nf-core/modules, so I am not very familiar.

I will remove the trailling lines and push another commit to this PR and see what happens.

Comment (fmalmeida) [reply modules/raven/main.nf]: I removed and it seems to have solved

**Review by mahesh-panchal — DISMISSED** [review]

**Comment by fmalmeida** [comment]
I thought they were. When I created it locally and let it overwrite the previous one (created when the PR was open) the tests ran ok.

But the pytest went south.

Maybe it did not overwrite or wasn't pushed in the commit, but probably the problem is with the `test.yml` as pointed out.

I'll try to address this tomorrow.

**Comment by fmalmeida** [comment]
I changed it as a draft to try to understand it before running more tests ...

When I run it locally, it seems to work:

```pytest
PROFILE=docker pytest --tag raven --symlink --keep-workflow-wd
=================================================== test session starts ====================================================
platform linux -- Python 3.7.6, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/falmeida/Documents/GitHub/modules
plugins: anyio-3.5.0, workflow-1.6.0
collecting ... 
collected 601 items                                                                                                        

raven test_raven:
	command:   nextflow run ./tests/modules/raven -entry test_raven -c ./tests/config/nextflow.config -c ./tests/modules/raven/nextflow.config
	directory: /tmp/pytest_workflow_xsm2x94e/raven_test_raven
	stdout:    /tmp/pytest_workflow_xsm2x94e/raven_test_raven/log.out
	stderr:    /tmp/pytest_workflow_xsm2x94e/raven_test_raven/log.err
'raven test_raven' done.

tests/test_versions_yml.py sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 14%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 34%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 53%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 72%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 92%]
ssssssssssss.ssssssssssssssssssssssssssss                                                                            [ 98%]
tests/modules/raven/test.yml .......                                                                                 [100%] Keeping temporary directories and logs. Use '--kwd' or '--keep-workflow-wd' to disable this behaviour.


===================================================== warnings summary =====================================================
../../../miniconda3/lib/python3.7/site-packages/pytest_workflow/plugin.py:392
  /home/falmeida/miniconda3/lib/python3.7/site-packages/pytest_workflow/plugin.py:392: UserWarning: .git dir detected: /home/falmeida/Documents/GitHub/modules/.git. pytest-workflow will copy the entire .git directory and all files ignored by git. It is recommended to use the --git-aware option.
    f".git dir detected: {str(git_dir)}. pytest-workflow "

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================== 8 passed, 593 skipped, 1 warning in 377.83s (0:06:17) ===================================
```

and also, the outputs are generated:

```bash
$ ls /tmp/pytest_workflow_xsm2x94e/raven_test_raven/work/15/2305eb38f3f595fdb200b138e6f5fa
raven.cereal  test.fasta.gz  test.fastq.gz  test.gfa  test.gfa.gz  versions.yml
```

**Comment by mahesh-panchal** [comment]
You can also try in a different computing environment if you want (click the link):

https://gitpod.io/#https://github.com/nf-core/modules/pull/1087

If a terminal is not open, click on the three bars in the top left and open a terminal from there.

**Comment by fmalmeida** [comment]
Thanks for this tip. Using the link you've provided I was able to track the problem ... In the `test.yml` the outputs were pointed to the tmp dir created while creating the yml locally, and not to the `output/` folder.

Now the only error is with the conda profile which was expected from what [Michael has commented](https://github.com/nf-core/modules/pull/1087#pullrequestreview-810697049).

It seems that one library `libstdcxx-ng` is only available for linux (https://anaconda.org/conda-forge/libstdcxx-ng) and not all OS

**Review by mahesh-panchal — APPROVED** [review]

**Pull request merged**